tools/compileProtos.ts

```diff
@@ -0,0 +1,163 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import * as fs from 'fs';
+import * as path from 'path';
+import * as util from 'util';
+import * as pbjs from 'protobufjs/cli/pbjs';
+
+const readdir = util.promisify(fs.readdir);
+const readFile = util.promisify(fs.readFile);
+const stat = util.promisify(fs.stat);
+const pbjsMain = util.promisify(pbjs.main);
+
+const PROTO_LIST_REGEX = /_proto_list.json$/;
+
+/**
+ * Recursively scans directories starting from `directory` and finds all files
+ * matching `PROTO_LIST_REGEX`.
+ *
+ * @param {string} directory Path to start the scan from.
+ * @return {Promise<string[]} Resolves to an array of strings, each element is a full path to a matching file.
+ */
+async function findProtoJsonFiles(directory: string): Promise<string[]> {
+  const result: string[] = [];
+  const files = await readdir(directory);
+  for (const file of files) {
+    const fullPath = path.join(directory, file);
+    const fileStat = await stat(fullPath);
+    if (fileStat.isFile() && file.match(PROTO_LIST_REGEX)) {
+      result.push(fullPath);
+    } else if (fileStat.isDirectory()) {
+      const nested = await findProtoJsonFiles(fullPath);
+      result.push(...nested);
+    }
+  }
+  return result;
+}
+
+/**
+ * Normalizes the Linux path for the current operating system.
+ *
+ * @param {string} filePath Linux-style path (with forward slashes)
+ * @return {string} Normalized path.
+ */
+function normalizePath(filePath: string): string {
+  return path.join(...filePath.split('/'));
+}
+
+/**
+ * Returns a combined list of proto files listed in all JSON files given.
+ *
+ * @param {string[]} protoJsonFiles List of JSON files to parse
+ * @return {Promise<string[]>} Resolves to an array of proto files.
+ */
+async function buildListOfProtos(protoJsonFiles: string[]): Promise<string[]> {
+  const result: string[] = [];
+  for (const file of protoJsonFiles) {
+    const directory = path.dirname(file);
+    const content = await readFile(file);
+    const list = JSON.parse(content.toString()).map((filePath: string) =>
+      path.join(directory, normalizePath(filePath))
+    );
+    result.push(...list);
+  }
+  return result;
+}
+
+/**
+ * Runs `pbjs` to compile the given proto files, placing the result into
+ * `./protos/protos.json`.
+ *
+ * @param {string[]} protos List of proto files to compile.
+ */
+async function compileProtos(protos: string[]): Promise<void> {
+  const pbjsArgs = [
+    '--target',
+    'json',
+    '-p',
+    path.join('node_modules', 'google-gax', 'protos'),
```

Comment:
does `google-gax` export this as a `bin`? it's better to execute `./node_modules/.bin/protos` if possible; because this will work regardless of hoisting.

Reply:
This is a value for `-p` option of `pbjs`, it basically points to an output directory, it's not executing anything.